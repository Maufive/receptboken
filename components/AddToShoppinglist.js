import React, { Component } from "react";
import Modal from "react-modal";
import axios from "axios";
import NotesIcon from "../svg/notes.svg";
import AddIcon from "../svg/add.svg";
import { Button } from "./styles/Button";
import { customStyles } from "./styles/Jumbotron";
import { LoginForm } from "./styles/ModalStyles";

Modal.setAppElement("#__next");

class AddToShoppinglist extends Component {
	state = {
		isModalOpen: false,
		title: "",
		shoppinglists: null
	};

	componentDidMount = () => {
		this.setState({ isModalOpen: this.props.open });
	};

	openModal = () => {
		this.setState({ isModalOpen: true });
		this.getShoppinglists();
	};

	closeModal = () => {
		this.setState({ isModalOpen: false });
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	onSubmit = async e => {
		e.preventDefault();
		const { title } = this.state;
		axios
			.post(`http://localhost:7777/lists/add/${this.props.receptid}`, { title })
			.then(response => {
				this.getShoppinglists();
			})
			.catch(error => console.log(error));
	};

	getShoppinglists = async () => {
		axios
			.get(`http://localhost:7777/lists/user/${this.props.userid}`)
			.then(response => this.setState({ shoppinglists: response.data }))
			.catch(error => console.log(error));
	};

	render() {
		const { shoppinglists } = this.state;
		return (
			<div>
				<Button onClick={this.openModal}>
					<NotesIcon /> Lägg till i inköpslista
				</Button>
				<Modal
					isOpen={this.state.isModalOpen}
					onAfterOpen={this.afterOpenModal}
					onRequestClose={this.closeModal}
					style={customStyles}
					contentLabel="Login Modal"
				>
					<LoginForm onSubmit={this.onSubmit}>
						<h2>Välj inköpslista</h2>
						<div>
							<input
								type="text"
								value={this.state.title}
								name="title"
								onChange={this.saveToState}
								required
							/>
							<label htmlFor="title">
								<AddIcon />
								Ny inköpslista
							</label>
						</div>
						<Button fullWidth type="submit">
							Skapa ny inköpslista
						</Button>
					</LoginForm>
					{shoppinglists &&
						shoppinglists.length >= 1 &&
						shoppinglists.map(list => <p>{list.title}</p>)}
				</Modal>
			</div>
		);
	}
}

export default AddToShoppinglist;
