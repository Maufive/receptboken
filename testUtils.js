export const fakeRecept = {
	_id: "1234",
	author: "5678436374re67fwuf76",
	created: "Idag",
	description: ["Steg1", "Steg2", "Steg3"],
	id: "1337",
	ingredients: [
		{
			input: "Lök",
			numberOfUnits: "2",
			units: "dl"
		}
	],
	photo: "asdf123.jpg",
	reviews: [{ score: 1 }],
	servings: "2",
	tags: ["Mat", "Gott"],
	timeRequired: "20 min",
	title: "Mitt jättegoda recept"
};

export const fakeUser = {
	_id: "1234",
	username: "fakeuser@gmail.com",
	password: "1234",
	hearts: [],
	fname: "Niklas",
	lname: "Albinsson",
	description: "Hej jag heter Niklas",
	photo: "http://fakephoto.jpg"
};

export const fakeShoppinglist = [
	{
		_id: "34567876543452176",
		title: "My Shoppinglist",
		created: "12/3 2019",
		author: "61386712376576d7q6ewd65",
		list: [
			{
				input: "Lök",
				numberOfUnits: "5",
				units: "dl"
			},
			{
				input: "Tomater",
				numberOfUnits: "10",
				units: "st"
			}
		]
	}
];
