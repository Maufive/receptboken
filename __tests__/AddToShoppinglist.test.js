import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import AddToShoppinglist from "../components/AddToShoppinglist";
import { fakeShoppinglist } from "../testUtils";

describe("<AddToShoppinglist />", () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<AddToShoppinglist />)));

	it("renders to the page and matches the snapshot", () => {
		expect(toJSON(wrapper)).toMatchSnapshot();
	});

	// Stuff to test:
	// Button #open-shoppinglist-modal
	it("opens the modal when the open-modal-button is clicked", () => {
		wrapper.find("#open-shoppinglist-modal").simulate("click");
		wrapper.update();
		expect(wrapper.state("isModalOpen")).toBe(true);
	});

	// Input #title
	describe("Title input", () => {
		it("should respond to change to the input and update the component", () => {
			wrapper.find("#title").simulate("change", {
				target: { name: "title", value: "Testing shoppinglist" }
			});

			expect(wrapper.state("title")).toEqual("Testing shoppinglist");
		});
	});

	it("renders a list of shoppinglists if the user has any", async () => {
		const Component = shallow(<AddToShoppinglist />);
		Component.setState({ isModalOpen: true, shoppinglists: fakeShoppinglist });
		await Component.update();
		expect(Component.find("#shoppinglists").children().length).toBe(1);
	});

	// When there are 1 or more lists in state, make sure they are rendered to the page
	it("does not render shoppinglists if there arent any", () => {
		expect(wrapper.find("List").children().length).toBe(0);
	});
});
