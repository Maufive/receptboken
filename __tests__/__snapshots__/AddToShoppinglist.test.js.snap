// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToShoppinglist /> renders to the page and matches the snapshot 1`] = `
<div>
  <Button
    id="open-shoppinglist-modal"
    onClick={[Function]}
    primary={true}
  >
    <ListBoldIcon
      viewBox="0 0 60 60"
      xmlns="http://www.w3.org/2000/svg"
    />
     Lägg till i inköpslista
  </Button>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Login Modal"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "background": "#fff",
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "zIndex": 10,
        },
      }
    }
  >
    <ModalStyles__LoginForm
      onSubmit={[Function]}
    >
      <h2>
        Välj inköpslista
      </h2>
      <div>
        <input
          id="title"
          name="title"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <label
          for="title"
          htmlFor="title"
        >
          <AddIcon
            viewBox="0 0 42 42"
            xmlns="http://www.w3.org/2000/svg"
          />
          Ny inköpslista
        </label>
      </div>
      <Button
        fullWidth={true}
        id="submit-shoppinglist"
        primary={true}
        type="submit"
      >
        Skapa ny inköpslista
      </Button>
      <ShoppinglistStyles__List
        id="shoppinglists"
      />
    </ModalStyles__LoginForm>
  </Modal>
</div>
`;
